### ISSUE TITLE
JasperReports Integration and QR code throws exception

### ISSUE THREAD

**Issue by tiagosteil** [daust/JasperReportsIntegration#27]
Hi, I'm trying to print a report with QR Code, but I'm not getting it.
I am using JasperReportsIntegration-2.4.0.0 with Apache Tomcat / 7.0.76 and Oracle Apex 19.1.

In the preview of Jaspersoft Studio the QR Code is displayed correctly, however when I copy the report to the server and run via apex, it shows the error below:

exception

javax.servlet.ServletException: Servlet execution threw an exception
	org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
root cause

java.lang.NoClassDefFoundError: org/apache/xml/serializer/TreeWalker
	org.apache.xalan.processor.TransformerFactoryImpl.newTransformer(TransformerFactoryImpl.java:818)
	net.sf.jasperreports.components.barcode4j.QRCodeSVGImageProducer.createImage(QRCodeSVGImageProducer.java:171)
	net.sf.jasperreports.components.barcode4j.AbstractBarcodeEvaluator.evaluateBarcodeRenderable(AbstractBarcodeEvaluator.java:109)
	net.sf.jasperreports.components.barcode4j.AbstractBarcodeEvaluator.visitQRCode(AbstractBarcodeEvaluator.java:516)
	net.sf.jasperreports.components.barcode4j.QRCodeComponent.receive(QRCodeComponent.java:49)
	net.sf.jasperreports.components.barcode4j.AbstractBarcodeEvaluator.evaluateBarcode(AbstractBarcodeEvaluator.java:85)
	net.sf.jasperreports.components.barcode4j.BarcodeFillComponent.evaluateBarcode(BarcodeFillComponent.java:83)
	net.sf.jasperreports.components.barcode4j.BarcodeFillComponent.evaluate(BarcodeFillComponent.java:70)
	net.sf.jasperreports.engine.fill.JRFillComponentElement.evaluate(JRFillComponentElement.java:110)
	net.sf.jasperreports.engine.fill.JRFillElementContainer.evaluate(JRFillElementContainer.java:381)
	net.sf.jasperreports.engine.fill.JRFillBand.evaluate(JRFillBand.java:500)
	net.sf.jasperreports.engine.fill.JRVerticalFiller.fillColumnBand(JRVerticalFiller.java:2022)
	net.sf.jasperreports.engine.fill.JRVerticalFiller.fillDetail(JRVerticalFiller.java:748)
	net.sf.jasperreports.engine.fill.JRVerticalFiller.fillReportStart(JRVerticalFiller.java:255)
	net.sf.jasperreports.engine.fill.JRVerticalFiller.fillReport(JRVerticalFiller.java:115)
	net.sf.jasperreports.engine.fill.JRBaseFiller.fill(JRBaseFiller.java:582)
	net.sf.jasperreports.engine.fill.BaseReportFiller.fill(BaseReportFiller.java:396)
	net.sf.jasperreports.engine.fill.JRFiller.fill(JRFiller.java:90)
	net.sf.jasperreports.engine.JasperFillManager.fill(JasperFillManager.java:286)
	de.oc.integration.jasper.webapp.ReportWrapper.service(Unknown Source)
	javax.servlet.http.HttpServlet.service(HttpServlet.java:731)
	org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
root cause

java.lang.ClassNotFoundException: org.apache.xml.serializer.TreeWalker
	org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1892)
	org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1735)
	org.apache.xalan.processor.TransformerFactoryImpl.newTransformer(TransformerFactoryImpl.java:818)
	net.sf.jasperreports.components.barcode4j.QRCodeSVGImageProducer.createImage(QRCodeSVGImageProducer.java:171)
	net.sf.jasperreports.components.barcode4j.AbstractBarcodeEvaluator.evaluateBarcodeRenderable(AbstractBarcodeEvaluator.java:109)
	net.sf.jasperreports.components.barcode4j.AbstractBarcodeEvaluator.visitQRCode(AbstractBarcodeEvaluator.java:516)
	net.sf.jasperreports.components.barcode4j.QRCodeComponent.receive(QRCodeComponent.java:49)
	net.sf.jasperreports.components.barcode4j.AbstractBarcodeEvaluator.evaluateBarcode(AbstractBarcodeEvaluator.java:85)
	net.sf.jasperreports.components.barcode4j.BarcodeFillComponent.evaluateBarcode(BarcodeFillComponent.java:83)
	net.sf.jasperreports.components.barcode4j.BarcodeFillComponent.evaluate(BarcodeFillComponent.java:70)
	net.sf.jasperreports.engine.fill.JRFillComponentElement.evaluate(JRFillComponentElement.java:110)
	net.sf.jasperreports.engine.fill.JRFillElementContainer.evaluate(JRFillElementContainer.java:381)
	net.sf.jasperreports.engine.fill.JRFillBand.evaluate(JRFillBand.java:500)
	net.sf.jasperreports.engine.fill.JRVerticalFiller.fillColumnBand(JRVerticalFiller.java:2022)
	net.sf.jasperreports.engine.fill.JRVerticalFiller.fillDetail(JRVerticalFiller.java:748)
	net.sf.jasperreports.engine.fill.JRVerticalFiller.fillReportStart(JRVerticalFiller.java:255)
	net.sf.jasperreports.engine.fill.JRVerticalFiller.fillReport(JRVerticalFiller.java:115)
	net.sf.jasperreports.engine.fill.JRBaseFiller.fill(JRBaseFiller.java:582)
	net.sf.jasperreports.engine.fill.BaseReportFiller.fill(BaseReportFiller.java:396)
	net.sf.jasperreports.engine.fill.JRFiller.fill(JRFiller.java:90)
	net.sf.jasperreports.engine.JasperFillManager.fill(JasperFillManager.java:286)
	de.oc.integration.jasper.webapp.ReportWrapper.service(Unknown Source)
	javax.servlet.http.HttpServlet.service(HttpServlet.java:731)
	org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)

**Comment by tiagosteil** [comment]
Error print:
What can I do ?

<img width="805" alt="erro" src="https://user-images.githubusercontent.com/27566133/81760077-b7f02880-949c-11ea-8629-4f4976313abe.png">

**Comment by tiagosteil** [comment]
Hi,
After some research I managed to successfully print my precious QR Code!

To achieve this I did the following steps:
1 - Download and copy de file `serializer-2.7.1.jar` to the folder:
_/var/lib/tomcat/webapps/JasperReportsIntegration/WEB-INF/lib_ on the server.

2 - Edit the file _/var/lib/tomcat/webapps/JasperReportsIntegration/WEB-INF/classes/jasperreports.properties_ and added the line:
`net.sf.jasperreports.components.barcode4j.image.producer=image`

3 - DONE, the QR Code appears in the report!

I hope this can help someone in the future.

Sources: https://stackoverflow.com/questions/25200599/tomcat-after-undeploy-deploy-a-war-makes-noclassdeffounderror-org-apache-xml-se and https://stackoverflow.com/questions/36298271/jasperserver-qrcode-not-visible-in-generated-report/36400420#36400420

**Comment by daust** [comment]
Great point @tiagosteil. 
This issue has been raised recently again in issue #55 . 

I have included the ``serializer-2.7.1.jar`` into the ``jri.war`` file with the gradle commands: 
```
    // https://mvnrepository.com/artifact/xalan/serializer
    // see bug: #55 - Barcodes produce classnotfounderror
    compile group: 'xalan', name: 'serializer', version: '2.7.1'
    // https://mvnrepository.com/artifact/org.apache.xmlgraphics/xmlgraphics-commons
    // see bug: #55 - Barcodes produce classnotfounderror
    compile group: 'org.apache.xmlgraphics', name: 'xmlgraphics-commons', version: '2.+'
```

Actually, I needed both the serializer and the xmlgraphics libraries. I was hitting another missing class error. 
The change to the ``jasperreports.properties`` file is no longer required as of jasperreports 6.1, see here: 
https://community.jaspersoft.com/jasperreports-server/issues/4911#comment-816311 

So, this is issue is now fixed in JasperReportsIntegration release 2.6.2 through issue #55 . 
Thanks
~Dietmar.
